Cookie Store API: add branch for empty string path

annevk opened the pull request:
#### 2c5cf3d08ebf52730f55b3ff867213971049cb29
<pre>
Cookie Store API: add branch for empty string path
<a href="https://bugs.webkit.org/show_bug.cgi?id=297268">https://bugs.webkit.org/show_bug.cgi?id=297268</a>

Reviewed by Rupin Mittal.

Implement <a href="https://github.com/whatwg/cookiestore/pull/283">https://github.com/whatwg/cookiestore/pull/283</a> with tests
upstreamed at <a href="https://github.com/web-platform-tests/wpt/pull/54264">https://github.com/web-platform-tests/wpt/pull/54264</a>

Canonical link: <a href="https://commits.webkit.org/298681@main">https://commits.webkit.org/298681@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/91f4b2b3229d0b5ae08e1968b980062b2d9a64dd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116313 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35974 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26521 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122370 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66874 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f7e08405-b25a-421c-ad1c-5f8869ec1f96) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118202 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36668 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44562 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88344 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42843 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/672d3b76-ee61-4a79-8201-b9416b4f7fe6) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119262 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29244 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104354 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68773 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28326 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22455 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66051 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98624 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22615 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125519 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43207 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32444 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97054 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43572 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100559 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96848 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42140 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20030 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/39157 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18577 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43094 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48686 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42561 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45896 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44266 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

RupinMittal commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -464,18 +464,13 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
+        return;
     }
 
+    // FIXME: The specification does not perform this initialization of domain.
```
I think this code is correct and it's actually the spec that is wrong here. It says here https://cookiestore.spec.whatwg.org/#CookieStore-set (in the green box) that the domain defaults to "Domain: same as the domain of the current document or service worker’s location"

Yet the steps say "Let domain be null". I'm not sure how the spec expects the domain to be default to the same as that of the current document if it defaults to null in the steps. 

I think this is a mistake in the spec maybe? Because last year, the default for domain was null (WebKit always defaulted to current document domain since CFNetwork won't let us set cookies with null domain). At some point the spec default was changed to current document domain but maybe they forgot to change in the steps?

(Contrast this with path, whose default is "/" and steps indeed start it off with "/"). Of course, the default is being changed in this patch here, but still the spec will be consistent in both places for path. It looks like a change needs to be made for domain to be consistent.

annevk replied:
That's what the network layer ends up defaulting domain to. The domintro box is arguably a bit misleading, but it's also non-normative and not aimed at implementers.

RupinMittal replied:
Right, the comment is ok. I do remember that when implementing, I had to add this even though the spec didn't have it because setting cookies simply doesn't work without it (due to our network layer). Hence that one bug I filed. Then at some point I saw that the domintro box was changed, so I thought that the spec had in fact changed from default null to the current context's domain. And I thought great! Our quirk is correct now. 

But I'm now realizing from your comment that's not fully true, and we're still figuring out that area. So yes, the comment is ok for now.

RupinMittal commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -464,18 +464,13 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
-            promise->reject(Exception { ExceptionCode::TypeError, "If the cookie name begins with \"__Host-\", the path must either not be specified or be \"/\"."_s });
-            return;
-        }
+    // FIXME: This should be further down.
```
It seems like the spec should figure out the steps of where it wants to set the default domain (see my below comment), which will tell us where this should go.

annevk replied:
It's a separate problem from what this PR is tackling though. I was hoping to tackle that next, though I think the changes will be mostly on our side.

RupinMittal replied:
Sounds good. Yeah since you're working on this area, I think the FIXME is ok for now then.

RupinMittal commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -507,17 +502,24 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
-            promise->reject(Exception { ExceptionCode::TypeError, "The path must begin with a '/'"_s });
-            return;
-        }
+    ASSERT(!cookie.path.isNull());
```
It's not possible for this to happen. `CookieInit::path` is default initialized to "/". Perhaps we can remove this assertion.

annevk replied:
Happy to remove it. The previous code checked for it being null so I thought I'd clarify that's not possible.

RupinMittal replied:
Actually yeah good point. It also doesn't hurt to have that there. Also a good safeguard just in case we change the default for path later too. 

Let's keep it.

RupinMittal commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -507,17 +502,24 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
+    }
+
+    if (cookie.path != "/"_s && cookie.name.startsWithIgnoringASCIICase("__Host-"_s)) {
+        promise->reject(Exception { ExceptionCode::TypeError, "If the cookie name begins with \"__Host-\", the eventual path must be \"/\"."_s });
```
Nit: What's the point of "eventual" here? Perhaps it's most clear to simply say `"... the path must be "/".`

RupinMittal replied:
Actually I get it. You're trying to explain the whole "it should either be "/" or empty so it can default to that" in a nice way.

Maybe we can say `... the path must be either be specified as "/" or not be specified so it defaults to "/".` to make it clear to the developer.

RupinMittal reviewed (approved):
Looks great :)

annevk committed:
Cookie Store API: add branch for empty string path
https://bugs.webkit.org/show_bug.cgi?id=297268

Reviewed by Rupin Mittal.

Implement https://github.com/whatwg/cookiestore/pull/283 with tests
upstreamed at https://github.com/web-platform-tests/wpt/pull/54264

Canonical link: https://commits.webkit.org/298681@main
